app/controllers/brexit_feedback_controller.rb

```diff
@@ -0,0 +1,58 @@
+require 'date'
+
+require 'google/apis/drive_v2'
+require 'google/api_client/client_secrets'
+
+require 'signet/oauth_2/client'
+
+class BrexitFeedbackController < ApplicationController
+  def home
+    session.destroy
+  end
+
+  def results
+    client_hash = session['auth_client']
+
+    opts = {
+            authorization_uri: "#{client_hash['authorization_uri']['scheme']}://#{client_hash['authorization_uri']['host']}#{client_hash['authorization_uri']['path']}",
+            token_credential_uri: "#{client_hash['token_credential_uri']['scheme']}://#{client_hash['token_credential_uri']['host']}#{client_hash['token_credential_uri']['path']}",
+            client_id: client_hash['client_id'],
+            client_secret: client_hash['client_secret'],
+            scope: client_hash["scope"][0],
+            redirect_uri: "#{client_hash['redirect_uri']['scheme']}://#{client_hash['redirect_uri']['host']}#{client_hash['redirect_uri']['path']}"
+           }
+
+    auth_client = Signet::OAuth2::Client.new(opts)
+
+    auth_client.code = params['code']
+    auth_client.additional_parameters = { access_type: 'offline', include_granted_scopes: 'true' }
+    brexit_slugs = Support::Requests::BrexitSlugFetcher.new(auth_client).slugs
+    @results = Support::Requests::BrexitFeedbackRequest.new(Date.parse(session['from_date']), Date.parse(session['to_date']), brexit_slugs).formatted_results
+  end
+
+  def auth
+    secrets = JSON.parse(File.read('client_secrets.json'))
+    secrets["web"]["client_id"] = ENV["google_client_id"]
+    secrets["web"]["client_secret"] = ENV["google_client_secret"]
+    File.open("client_secrets.json", "w+") do |f|
```

Comment:
Does this `#auth` method get called for every request? It's doing IO repeatedly on the same file, I wonder if there's a more efficient way of handling this. Can the authorisation be stored in memory once generated?

Reply:
So as it stands the auth method is being called for every request. I've moved things around a bit (pr up soon) so that there's one static file that holds a json.erb template and a method that writes to a new file in the root directory. I'm not too bothered by the fact that the authorisation isn't saved because the use case for this is one person running it once a week. If this changes I'd be happy to look into making this a little smoother. Do you know if there's an elegant way to delete a file after a page has been rendered?

Reply:
I think reading and writing a file for every request is a limited approach, additionally deleting it will only add to the I/O.
It might be better to extend how `Google::APIClient::ClientSecrets.load` works to do what we want here.
If you [look at the source](https://github.com/googleapis/google-api-ruby-client/blob/master/lib/google/api_client/client_secrets.rb#L73) for `Google::APIClient::ClientSecrets.load`  it returns the result of `data = File.open(filename, 'r') { |file| JSON.load(file.read) }` We could extend this class it a load method of our own and make the process much tidier.